docs/explanation/holistic-vs-delta-charms.md

```diff
@@ -1,66 +1,230 @@
-        try:
-            self._update_redis_config(container, redis_port)
-        except ops.pebble.ConnectionError:
-            # config-changed happened first, wait for pebble-ready
-            return
+### Example
+
+```py
+def __init__(self, framework: ops.Framework):
+    super().__init__(framework)
+    self.framework.observe(self.on.install, self._on_install)
+    self.framework.observe(self.on.start, self._on_start)
+
+    hostname = socket.getfqdn()
+    self.foo = FooRequires(self, "foo-relation", address=hostname)
+    self.framework.observe(self.foo.on.data_available, self._on_data_available)
+
+    self.bar = BarProvider(self, "bar-relation")
+    self.framework.observe(self.bar.on.create_bar, self._on_create_bar)
+
+def _on_install(self, event: ops.InstallEvent):
+    # Install the workload binary
+    apt.add_package("some-workload")
+
+def _on_start(self, event: ops.StartEvent):
+    # This unit has been started, peer relations are usable
+    if self.unit.is_leader():
+        ...
+
+def _on_data_available(self, event: DataAvailableEvent):
+    # Update the workload with event's data
+    self.workload.reconfigure(some_key=event.some_value)
+
+def _on_create_bar(self, event: CreateBarEvent):
+    # Provision a `Bar` resourse in the workload
+    self.workload.create_foo(event.some_field)
 ```
 
+Notice how the handler methods map to Juju and custom events directly.
+There's less boilerplate and flow control is predictable.
+
+### Developing your charm
+
+Unit testing is initially straightforward, however both charm and tests get messy when event interdependencies are accounted for:
+
+- creating a resource is only possible after the workload has been installed
+- the available data typically needs to be mixed with application's configuration
+- cases where something is inferred from sequence of events [Juju being helpful](https://discourse.charmhub.io/t/charm-lifecycle/5938)
+
+A good rule of thumb is this: if you're starting to use `defer` in various places, consider whether it's time to rewrite the charm using the reconciler pattern.
+
+## Which approach to use?
+
+At a high level, simple workloads are served well by delta charms.
+Complex workloads and especially mature feature-rich charms utilising many charm libraries are more robust if the reconciler pattern is followed.
+
+### Dedicated handlers for specific Juju events
+
+Some events are special, and are typically processed outside the reconciler pattern in holistic charms.
+In both styles of charms a dedicated event handler is used for:
+
+- unit lifecycle end events, because the goal of reconciliation is different
+- action event, as it is synchronous and the event object holds arguments and result sink
+- secret rotation event, where secret must be modified synchronously
+- secret remove event, because the revision that needs removal is not available otherwise
+- Ops lifecycle events, like `update-unit-status`, which accompany the Juju event
```

Comment:
Probably meant to be `collect-unit-status`?

Reply:
like `on_collect_unit_status`